### PR TITLE
llvm: general cleanup and corrections

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Some of this could likely be applied to llvm@9 and earlier. The formula was a collection of a years of additions - some of which no longer applies.

~~I will want to verify the bottle's contents before merging this.~~ Done.

### CODE REVIEW

File: Formula/llvm.rb [hunk 1]
```diff
@@ -113,22 +74,37 @@ class Llvm < Formula
...
+    unless build.head?
+      llvmpath.install buildpath.children - [buildpath/".brew_home"]
+      (projects + runtimes).each { |p| resource(p).stage(buildpath/p) }
+    end
+
+    # Needed until https://reviews.llvm.org/D63883 lands again.
+    # Use system libcxxabi.
+    rm_r "libcxxabi" if build.head?
+
+    py_ver = "2.7"
```
Comment (SMillerDev) [inline Formula/llvm.rb]: 3 isn't an option?

**Review by SMillerDev — COMMENTED** [review]
Should this also be applied to older versions?

**Comment by Bo98** [comment]
> Should this also be applied to older versions?

Perhaps in part - I need to be careful.

File: Formula/llvm.rb (hunk 1 above)
Comment (Bo98) [reply Formula/llvm.rb]: I think it is - Linuxbrew uses it. I limited making any actual functional changes (besides fixing polly linking) but I can go ahead and migrate to 3.8 unless there is a particular reason anyone knows not to.

Comment (iMichka) [reply Formula/llvm.rb]: If we can use python 3.8, I would prefer that too. Just had no time to migrate it until now.

Comment (Bo98) [reply Formula/llvm.rb]: Apart from Python, is there anything else here you would prefer done here that would be useful for Linux?

Comment (SMillerDev) [reply Formula/llvm.rb]: @iMichka I linked the monster patch for linux in a slack PM to Bo

Comment (Bo98) [reply Formula/llvm.rb]: I think I'll need to better understand the original problem and why it is Linux specific.

Looking at the patch alone, it doesn't seem obvious at first if there's a better way. You can do things like supplying sysroots but I think that is not what is wanted.

**Comment by Bo98** [comment]
Actually this doesn't need `depends_on "python@3.8"` at all really. It's not invoked at all at compile time and it makes no sense mandate it at runtime. I'll just install it to the correct place so that those who have `python@3.8` installed can use it.

**Comment by SMillerDev** [comment]
Sounds good

File: Formula/llvm.rb [hunk 2]
```diff
@@ -137,55 +113,47 @@ def install
...
-      tools/clang/tools/clang-format/clang-format.el
-      tools/clang/tools/clang-rename/clang-rename.el
-      tools/clang/tools/extra/clang-include-fixer/tool/clang-include-fixer.el
-    ]
+    # Install LLVM Python bindings
+    # Clang Python bindings are installed by CMake
+    (lib/"python#{py_ver}/site-packages").install llvmpath/"bindings/python/llvm"
+
+    # Install Emacs modes
+    elisp.install Dir[llvmpath/"utils/emacs/*.el"] + Dir[share/"clang/*.el"]
```
[the remark this answers is not in the document]
Comment (Bo98) [reply Formula/llvm.rb]: Yes, I downloaded the bottle from Jenkins to verify:

```
$ ls ./llvm/10.0.0/share/emacs/site-lisp/llvm
clang-format.el         clang-rename.el  llvm-mode.el
clang-include-fixer.el  emacs.el         tablegen-mode.el
```

**Comment by josephg** [comment]
I'm still running into polly linking issues using the formula in this PR. I'm trying to build [zig](https://github.com/ziglang/zig/) from source:

```
$  brew install https://raw.githubusercontent.com/Bo98/homebrew-core/llvm-cleanup/Formula/llvm.rb
(4 hours later)
zig/build$ cmake .. -DCMAKE_PREFIX_PATH=/usr/local/opt/llvm  -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++

Configuring zig version 0.5.0+f6d384450
...
-- Build files have been written to: /Users/josephg/3rdparty/zig/build
$ make
...
Undefined symbols for architecture x86_64:
  "getPollyPluginInfo()", referenced from:
      (anonymous namespace)::EmitAssemblyHelper::EmitAssemblyWithNewPassManager(clang::BackendAction, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream> >) in libclangCodeGen.a(BackendUtil.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [zig0] Error 1
make[1]: *** [CMakeFiles/zig0.dir/all] Error 2
make: *** [all] Error 2
```

Mangled symbols for getPollyPluginInfo show up in libclang-cpp.dylib and LLVMPolly.so:

```
$ nm -gU  /usr/local/opt/llvm/lib/libclang-cpp.dylib | grep getPolly                                    
0000000001d294f6 T __Z18getPollyPluginInfov
```

**Comment by Bo98** [comment]
The polly fix in this PR was a different one - where Polly isn’t linked into LLVM subtools.

That said, I thought the issue you describe was specific to standalone clang builds. The third commit of this pull request changes the type of build LLVM performs to avoid standalone builds entirely. If it doesn’t work on these two quite different setups, I don’t really know what can be done without upstream fixing it on their side.

**Comment by Bo98** [comment]
If someone can link to me a specific LLVM bug or patch about the issue that is not specific to Clang standalone then let me know. It seems to me that it's not a misconfiguration on Homebrew's side given apt.llvm.org has the same issue and there is a current open bug at Debian about it.

**Comment by jedisct1** [comment]
At least, this PR totally solved LLVM 10 installation for me.

**Comment by Bo98** [comment]
I see a lot of discussion pointing here from various issues so I'll summarise what's been done:

* This is now built using the "monorepo" layout. This is generally the layout that is the most tested upstream, and as such may well fix things on its own from changing to this. This also allows `--HEAD` builds to "work" again (beside from upstream bugs that may prevent compile).
* Fixed a compile issue if Lua was installed.
* Prevented opportunistic linkage to LZMA.
* Greatly simplified the formula to remove installs already handled by CMake and to reduce duplicated files and unneeded wrappers. No functionality has been lost.
* The binaries are now linked to the shared libraries rather than static. This decreases the size of the bottle and the install by 80%. No functionality has been lost, but let me know if there is any significant performance regressions (I doubt it since most other package managers ship it this way).
* libLLVM-C is now included.
* Python bindings are now supplied for Python 3.8 rather than 2.7.
* The tests have been improved to properly test both Xcode and CLT SDKs.
* Some old, broken Polly-related CMake variables were corrected. Probably no actual benefit here though since I suspect they were enabled by default anyway.

As for `zig`, I do not believe this is Homebrew specific at all. I've seen the same issue on at least 4 package managers so it seems to clear to me that it is an upstream LLVM bug. I recommend you speak with them - but feel free to let me know if any patch etc. is available.

I do suspect however that you can avoid the issue altogether by stopping linking against Clang's static libraries and start linking against the shared library `clang-cpp`. I know Fedora for one have made an effort to move everything they manage to use `clang-cpp`. I've not tried it in the context of `zig` however.

LLVM does provide some CMake targets for you in the CMake files they ship at `lib/cmake`. This is the recommended approach to linking LLVM & Clang in CMake. `llvm-config` doesn't cover `clang-cpp` as far as I know and has known bugs on macOS so I strongly discourage using it within CMake as much as possible. Based on some of the logs I've seen over the past couple days, those bugs are clearly still present.

**Comment by moreindirection** [comment]
Thanks for fixing - llvm install cleanly for me now.

**Comment by timotheecour** [comment]
@Bo98 that commit https://github.com/Homebrew/homebrew-core/commit/78baeec70f6db0cec41da6c1fbb2f1fd5cb2734f only does a small subset of the changes in this PR, what about the other changes?

according to https://github.com/ziglang/zig/issues/4897#issuecomment-607541324
> @robinsoncol Homebrew’s LLVM is misconfigured: Homebrew/homebrew-core#52352

**Comment by Bo98** [comment]
@timotheecour See discussion on https://github.com/ziglang/zig/issues/4799.

I'm aware of some issues on zig 0.6.0 but they don't appear to be the same issues as the one you linked.

**Comment by timotheecour** [comment]
i followed that thread but the "fix" suggested `-DZIG_PREFER_CLANG_CPP_DYLIB=ON` doesn't really work: it makes zig build but the zig is useless, see https://github.com/ziglang/zig/issues/5055#issuecomment-614548617 ; where i show that:
* with gcc + manually built llvm+friends + a patch to zig, zig works with or without DZIG_STATIC=ON
* with clang + homebrew built llvm+friends (or manually built), `-DZIG_PREFER_CLANG_CPP_DYLIB=ON` is needed to avoid the famous `getPollyPluginInfo` link error, but then zig is useless

**Pull request closed without merging**